llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py

```diff
@@ -0,0 +1,430 @@
+        callback_manager: Optional[CallbackManager] = None,
+        verbose: bool = False,
+        **kwargs: Any,
+    ) -> "SelfReflectionAgentWorker":
+        """Convenience constructor."""
+        if llm is None:
+            try:
+                from llama_index.llms.openai import OpenAI
+            except ImportError:
+                raise ImportError(
+                    "Missing OpenAI LLMs. Please run `pip install llama-index-llms-openai`."
+                )
+            llm = OpenAI(model="gpt-4-turbo-preview", temperature=0)
+
+        return cls(
+            llm=llm,
+            max_iterations=max_iterations,
+            callback_manager=callback_manager,
+            verbose=verbose,
+            **kwargs,
+        )
+
+    def initialize_step(self, task: Task, **kwargs: Any) -> TaskStep:
+        """Initialize step from task."""
+        # temporary memory for new messages
+        new_memory = ChatMemoryBuffer.from_defaults()
+
+        # put current history in new memory
+        messages = task.memory.get()
+        for message in messages:
+            new_memory.put(message)
+
+        # initialize task state
+        task_state = {
+            "new_memory": new_memory,
+            "sources": [],
+        }
+        task.extra_state.update(task_state)
+
+        return TaskStep(
+            task_id=task.task_id,
+            step_id=str(uuid.uuid4()),
+            input=task.input,
+            step_state={"count": 0},
+        )
+
+    @dispatcher.span
+    def _reflect(
+        self, chat_history: List[ChatMessage]
+    ) -> Tuple[Reflection, ChatMessage]:
+        """Reflect on the trajectory."""
+        reflection = self._reflection_program(
+            chat_history=messages_to_prompt(chat_history)
+        )
+
+        if self._verbose:
+            print(f"> Reflection: {reflection.dict()}")
+
+        # end state: return user message
+        reflection_output_str = (
+            f"Is Done: {reflection.is_done}\nCritique: {reflection.feedback}"
+        )
+        critique = REFLECTION_RESPONSE_TEMPLATE.format(
+            reflection_output=reflection_output_str
+        )
+
+        return reflection, ChatMessage.from_str(critique, role="user")
+
+    @dispatcher.span
+    def _correct(self, input_str: str, critique: str) -> ChatMessage:
+        correction = self._correction_program(input_str=input_str, feedback=critique)
+
+        correct_response_str = CORRECT_RESPONSE_FSTRING.format(
+            correction=correction.correction
+        )
+        if self._verbose:
+            print(f"Correction: {correction.correction}", flush=True)
+        return ChatMessage.from_str(correct_response_str, role="assistant")
+
+    @dispatcher.span
+    @trace_method("run_step")
+    def run_step(self, step: TaskStep, task: Task, **kwargs: Any) -> TaskStepOutput:
+        """Run step."""
+        state = step.step_state
+        state["count"] += 1
+
+        messages = task.extra_state["new_memory"].get()
```

Comment:
Does this include the initial task input? I might have missed where that gets added to memory

Reply:
yes it should, since the `introspective agent` first adds the chat history from the main agent which includews the initial task input to the reflective agent's memory.

https://github.com/run-llama/llama_index/blob/0d05f9090211a0c51a3796f052118fc5f720fc46/llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/step.py#L162